### PR TITLE
Notify users on groups

### PR DESCRIPTION
**Pull request by saraycp**
Follow-up PR #16225

Screenshot 2024-06-04 at 10-48-15 Open Build Service

### CODE REVIEW

File: src/api/app/models/event/added_user_to_group.rb
Comment (hennevogel) [inline src/api/app/models/event/added_user_to_group.rb:3]: Please don't send this over the message bus, it's a personal notification.

File: src/api/app/models/event/added_user_to_group.rb
Comment (hennevogel) [inline src/api/app/models/event/added_user_to_group.rb:10]:
```suggestion
      "The user #{payload['user']} has been added to the group #{payload['group']} by #{payload['who']}"
```

File: src/api/app/models/event/removed_user_from_group.rb
Comment (hennevogel) [inline src/api/app/models/event/removed_user_from_group.rb:3]: Please don't send this over the message bus, it's a personal notification.

File: src/api/app/models/event/removed_user_from_group.rb
Comment (hennevogel) [inline src/api/app/models/event/removed_user_from_group.rb:10]:
```suggestion
      "The user #{payload['user']} has been removed from to the group #{payload['group']} by #{payload['who']}"
```

**Review by hennevogel — CHANGES_REQUESTED** [review]

File: src/api/app/models/event/added_user_to_group.rb
Comment (saraycp) [reply src/api/app/models/event/added_user_to_group.rb]: I had rephrased the subject before I read your comment. It's finally: 

` "'#{payload['who']}' removed you from the group '#{payload['group']}'"`

File: src/api/app/models/event/removed_user_from_group.rb
Comment (saraycp) [reply src/api/app/models/event/removed_user_from_group.rb]: I had rephrased the subject before I read your comment. It's finally: 

` "'#{payload['who']}' removed you from the group '#{payload['group']}'"`

File: src/api/app/components/notification_component.rb
Comment (hennevogel) [inline src/api/app/components/notification_component.rb:14]: What is a "Group user"?

File: src/api/app/models/notification.rb
Comment (hennevogel) [inline src/api/app/models/notification.rb:40]: I doubt people want to filter for either. I would only make a `Group` scope/filter

File: src/api/app/components/notification_filter_component.rb
Comment (hennevogel) [inline src/api/app/components/notification_filter_component.rb:1]: This looks like a rebase problem. This class is gone in master...

File: src/api/app/models/notification.rb
Comment (saraycp) [reply src/api/app/models/notification.rb]: Unified into one.

File: src/api/app/components/notification_component.rb
Comment (saraycp) [reply src/api/app/components/notification_component.rb]: I changed it to `'GroupsUser' => 'User added/removed on a group'`

File: src/api/app/components/notification_filter_component.rb
Comment (saraycp) [reply src/api/app/components/notification_filter_component.rb]: Exactly. Removed already

File: src/api/app/components/notification_component.rb
Comment (krauselukas) [inline src/api/app/components/notification_component.rb]: @saraycp my bad, I forgot to add the `Event::RemovedUserFromGroup` here

File: src/api/app/models/group.rb
Comment (krauselukas) [inline src/api/app/models/group.rb:211]: @saraycp hmm, leaking the user session into the models is not great. I know we have many places where this is happening. But would be great if we could avoid this.

File: src/api/app/views/webui/users/notifications/_notification_filter.html.haml
Comment (krauselukas) [inline src/api/app/views/webui/users/notifications/_notification_filter.html.haml:74]: @saraycp The text looks still quite long in the filter area. Maybe we just call it `Group Roles`? or including it under the `Roles Granted` and `Roles Revoked` filters?

Comment (saraycp) [reply src/api/app/views/webui/users/notifications/_notification_filter.html.haml:74]: What about `User on Groups`? It's independent of the roles

File: src/api/app/components/notification_component.rb
Comment (saraycp) [reply src/api/app/components/notification_component.rb]: No problem

File: src/api/app/models/group.rb
Comment (saraycp) [reply src/api/app/models/group.rb:211]: :thinking: thinking of alternatives...

**Comment by danidoni** [comment]
Wouldn't it be better to split this PR into three? One for the creation of the event, another one for the delivery of the notification and the UI at the end?

**Comment by hennevogel** [comment]
> Wouldn't it be better to split this PR into three?

Because?

**Comment by danidoni** [comment]
> Because?

Because there is a lot of changes at once.

File: src/api/app/components/notification_notifiable_link_component.rb
Comment (hennevogel) [inline src/api/app/components/notification_notifiable_link_component.rb:45]:
```suggestion
      "#{event_payload['who'] || Someone} added you to the group '#{@notification.event_payload['group']}'"
```

File: src/api/app/components/notification_notifiable_link_component.rb
Comment (hennevogel) [inline src/api/app/components/notification_notifiable_link_component.rb:47]:
```suggestion
      "#{event_payload['who'] || Someone} added you to the group '#{@notification.event_payload['group']}'"
```

File: src/api/app/components/notification_notifiable_link_component.rb
Comment (hennevogel) [inline src/api/app/components/notification_notifiable_link_component.rb:118]: Link to the group show page?

File: src/api/app/models/event/added_user_to_group.rb
Comment (hennevogel) [inline src/api/app/models/event/added_user_to_group.rb:25]: Help me understand, why is the notifiable a `GroupUser` (the in-between model) and not `Group`?

**Review by hennevogel — CHANGES_REQUESTED** [review]

File: src/api/app/models/event/added_user_to_group.rb
Comment (saraycp) [reply src/api/app/models/event/added_user_to_group.rb:25]: Semantically, we notify about a new association between a Group and a User, that's GroupsUser.

In fact, I started using Group as notifiable, but I then thought it made more sense like this.

File: src/api/app/components/notification_notifiable_link_component.rb
Comment (saraycp) [reply src/api/app/components/notification_notifiable_link_component.rb:118]: The group show page is only accessible to admins and maintainers. It goes through this policy method:

https://github.com/openSUSE/open-build-service/blob/547b7fafa89b38ed22dc216f4d8dd31e2b70021e/src/api/app/policies/group_policy.rb#L12

So for the majority of users there won't be any link, so I decided not to add it. Admins and maintainers can easily reach the group in case they want to.

File: src/api/app/views/webui/users/notifications/_notification_filter.html.haml
Comment (saraycp) [reply src/api/app/views/webui/users/notifications/_notification_filter.html.haml]: Changed to `User on Groups`

File: src/api/app/components/notification_notifiable_link_component.rb
Comment (saraycp) [reply src/api/app/components/notification_notifiable_link_component.rb]: Changed

File: src/api/app/components/notification_notifiable_link_component.rb
Comment (saraycp) [reply src/api/app/components/notification_notifiable_link_component.rb]: Changed with "removed you from"

File: src/api/app/components/notification_notifiable_link_component.rb
Comment (saraycp) [reply src/api/app/components/notification_notifiable_link_component.rb:118]: You're right! I don't know what failed on my tests that I couldn't access and assumed the policy was the limitation.... :thinking: 

Alright, I'm going to add the link.

File: src/api/app/components/notification_notifiable_link_component.rb
Comment (saraycp) [reply src/api/app/components/notification_notifiable_link_component.rb]: This is the result:

Screenshot 2024-06-12 at 12-59-17 Open Build Service

File: src/api/app/components/notification_notifiable_link_component.rb
Comment (saraycp) [reply src/api/app/components/notification_notifiable_link_component.rb]: Done.

Screenshot 2024-06-12 at 13-08-19 Open Build Service

File: src/api/app/models/group.rb
Comment (krauselukas) [reply src/api/app/models/group.rb:211]: @saraycp alright, yeah, since this is always handled slightly different (API, webui, different methods) we would have to add a considerable amount of more code changes in order to avoid this. Its not nice, but dont want to block this from going further. Its maybe something for another refactoring. I will create a card to track it

File: src/api/app/components/notification_notifiable_link_component.rb
Comment (saraycp) [reply src/api/app/components/notification_notifiable_link_component.rb]: Same here, remove the description as it is duplicated code.
https://github.com/openSUSE/open-build-service/pull/16242#discussion_r1636248476

File: src/api/app/models/group.rb
Comment (saraycp) [reply src/api/app/models/group.rb:211]: We agreed on postponing this. It will require some refactoring. It's tracked on this [Trello card](https://trello.com/c/nw9Z76QI/183-stop-using-usersession-in-models)

File: src/api/app/models/event/added_user_to_group.rb
Comment (hennevogel) [reply src/api/app/models/event/added_user_to_group.rb:25]: Semantically we notify about a change to a `Group`.

Comment (saraycp) [reply src/api/app/models/event/added_user_to_group.rb:25]: I see. I'll change it.

Comment (saraycp) [reply src/api/app/models/event/added_user_to_group.rb]: Done!

File: src/api/app/controllers/webui/users/notifications_controller.rb
Comment (hennevogel) [inline src/api/app/controllers/webui/users/notifications_controller.rb:5]: Same here, there is on `user_on_groups` left.

File: src/api/app/models/event/removed_user_from_group.rb
Comment (hennevogel) [inline src/api/app/models/event/removed_user_from_group.rb:4]: Same here. `user` -> `member`

File: src/api/app/views/event_mailer/added_user_to_group.html.haml
Comment (hennevogel) [inline src/api/app/views/event_mailer/added_user_to_group.html.haml:1]: You should probably at least wrap this into a `<p>`...

File: src/api/spec/models/groups_user_spec.rb
Comment (hennevogel) [inline src/api/spec/models/groups_user_spec.rb:10]: How do you expect this to break?

**Review by hennevogel — CHANGES_REQUESTED** [review]

File: src/api/app/services/notification_service/web_channel.rb
Comment (danidoni) [inline src/api/app/services/notification_service/web_channel.rb:13]: I think we can get rid of this table. See https://trello.com/c/krml3hbT/184-remove-the-allowedfinders-from-notificationservice

File: src/api/app/views/event_mailer/added_user_to_group.html.haml
Comment (danidoni) [inline src/api/app/views/event_mailer/added_user_to_group.html.haml:1]: Why do we need the `"`?

File: src/api/spec/models/groups_user_spec.rb
Comment (danidoni) [inline src/api/spec/models/groups_user_spec.rb:7]: Using `login` in model specs is weird. Better wrap the code with`admin_user.run_as do ... end`

Comment (hennevogel) [reply src/api/spec/models/groups_user_spec.rb:7]: > Using login in model specs is weird

For sure not. We even have `ModelsAuthentication`...

File: src/api/app/components/notification_action_description_component.rb
Comment (saraycp) [reply src/api/app/components/notification_action_description_component.rb]: Same result. Condition skipped.

File: src/api/app/components/notification_component.rb
Comment (saraycp) [reply src/api/app/components/notification_component.rb]: Done!

File: src/api/app/models/event/added_user_to_group.rb
Comment (saraycp) [reply src/api/app/models/event/added_user_to_group.rb]: I like it! Changed!

File: src/api/app/models/event/added_user_to_group.rb
Comment (saraycp) [reply src/api/app/models/event/added_user_to_group.rb]: https://trello.com/c/mMCveMXQ/185-groups-maintainers-get-notified-when-a-user-is-added-to-the-group

File: src/api/app/controllers/person/notifications_controller.rb
Comment (saraycp) [reply src/api/app/controllers/person/notifications_controller.rb:7]: Indeed. Removed

File: src/api/app/views/event_mailer/added_user_to_group.erb
Comment (saraycp) [reply src/api/app/views/event_mailer/added_user_to_group.erb]: Corrected

File: src/api/app/views/webui/users/notifications/_notification_filter.html.haml
Comment (saraycp) [reply src/api/app/views/webui/users/notifications/_notification_filter.html.haml]: Done

File: src/api/app/controllers/webui/users/notifications_controller.rb
Comment (saraycp) [reply src/api/app/controllers/webui/users/notifications_controller.rb]: Done

File: src/api/app/views/event_mailer/added_user_to_group.html.haml
Comment (saraycp) [reply src/api/app/views/event_mailer/added_user_to_group.html.haml]: Done

File: src/api/spec/models/groups_user_spec.rb
Comment (saraycp) [reply src/api/spec/models/groups_user_spec.rb]: Sorry, I forgot to remove the specs from the previous PR.

File: src/api/app/services/notification_service/web_channel.rb
Comment (saraycp) [reply src/api/app/services/notification_service/web_channel.rb:13]: I have to double-check and take a decision.

File: src/api/app/views/event_mailer/removed_user_from_group.erb
Comment (saraycp) [reply src/api/app/views/event_mailer/removed_user_from_group.erb]: Corrected

File: src/api/app/services/notification_service/web_channel.rb
Comment (danidoni) [reply src/api/app/services/notification_service/web_channel.rb:13]: I mean, not now. I wrote a card to do it later...

File: src/api/app/views/webui/users/notifications/_notification_filter.html.haml
Comment (hennevogel) [inline src/api/app/views/webui/users/notifications/_notification_filter.html.haml:74]:
```suggestion
      = render partial: 'webui/shared/check_box', locals: { label: 'Group', amount: counted_notifications['member_on_groups'],
```

**Review by hennevogel — CHANGES_REQUESTED** [review]

File: src/api/app/services/notification_service/web_channel.rb
Comment (saraycp) [reply src/api/app/services/notification_service/web_channel.rb:13]: To start with, the finder was wrong because it used `@parameters['notifiable_id']` instead of `@parameters[:notifiable_id]`. There are other wrong finders, so I created this [Trello card](https://trello.com/c/UwjKoROV/186-revisit-outdatednotficationsfinder).

The goal of this finder is to avoid repeating the same type of notification many times. If someone adds and removes user U from group G many times, we keep only the last addition and removal. So I believe it makes sense to keep the functionality.

Please check the latest commit to see the fix.

File: src/api/app/queries/outdated_notifications_finder/group.rb
Comment (saraycp) [reply src/api/app/queries/outdated_notifications_finder/group.rb:1]: I wrote my conclusions here: https://github.com/openSUSE/open-build-service/pull/16242#discussion_r1638292113

File: src/api/app/views/webui/users/notifications/_notification_filter.html.haml
Comment (saraycp) [reply src/api/app/views/webui/users/notifications/_notification_filter.html.haml:74]: I didn't want to use `Group(s)` because there is a `Groups` section in the filter and could be confusing.

Screenshot(1)


I'll use `Groups membership` if you don't mind.

Comment (saraycp) [reply src/api/app/views/webui/users/notifications/_notification_filter.html.haml]: Done

**Comment by saraycp** [comment]
@danidoni @hennevogel I think all the feedback is addressed now.
I add a few examples on review-app: https://obs-reviewlab.opensuse.org/saraycp-notify_users_on_groups/my/notifications?state=unread&kind%5B%5D=member_on_groups

**Review by hennevogel — APPROVED** [review]

**Pull request merged**
